### PR TITLE
Gtplus merging grappa

### PR DESCRIPTION
**Pull request by gtdeveloper5**
Aim: to move grappa functionalities out of gtplus and move it into mri_core

1. add 2D and 3D grappa functions to mri_core
2. add mri_core_definition.h, moving definitions out of gtplus
3. add mri_core_utility to store useful utility functions
4. clean many functions from gtplus util

To do:

1. add a sum_over_dimension function for hoNDArray, replace all current sumOver functions with this general one
2. replace multipleMultiply with the one in hoNDArray_elemwise.h
3. move the coil map estimation functions from gtplus

### CODE REVIEW

**Comment by hansenms** [comment]
Thanks for getting started on this. 

It is nice that you have moved those definitions out of gtPlus, and it is nice that they have more reasonable names.

The points that you mention on your To Do have to be done, ie. those sum over dimensions, mutiply, etc. have to be replaced. I don't think it makes any sense to check this code in when large parts of it will be replaced? What were your thoughts? Did you want to merge first and then modify or continue to work in this branch before merging?

Also there are no test cases, so you need to make at least a couple of example gadgets that use the buffer structures with these functions so that it can be tested. Test cases must be added too (they can likely use existing data). 

I have not read through it all, but I don't understand why all the functions are wrapped in classes when they don't share information. What is the point of that organization?

More comments to follow and we should put it on the agenda to discuss Friday or whenever you are around again.

File: toolboxes/mri_core/mri_core_export.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef _MRI_CORE_EXPORT_H_
+#define _MRI_CORE_EXPORT_H_
+
+#if defined (WIN32)
+    #ifdef BUILD_TOOLBOX_STATIC
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_export.h]: I want this static toolbox stuff gone. It messes with a lot of stuff, specifically the ability to reliably count on singletons across platforms.

File: toolboxes/mri_core/mri_core_grappa.cpp [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+
+#ifdef USE_OMP
+    #include "omp.h"
+#endif // USE_OMP
+
+namespace Gadgetron
+{
+
+template <typename T> 
+void grappa<T>::printInfo(std::ostream& os)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_grappa.cpp]: What is this for?

File: toolboxes/mri_core/mri_core_grappa.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#include "ho4DArray.h"
+#include "ho5DArray.h"
+#include "ho6DArray.h"
+#include "ho7DArray.h"
+
+namespace Gadgetron
+{
+
+template <typename T> 
+class EXPORTMRICORE grappa : public parallelImaging<T>
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_grappa.h]: Don't understand why this is a class??

File: toolboxes/mri_core/mri_core_parallel_imaging.cpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+#ifdef USE_OMP
+    #include "omp.h"
+#endif // USE_OMP
+
+namespace Gadgetron
+{
+
+template <typename T> 
+void parallelImaging<T>::printInfo(std::ostream& os)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_parallel_imaging.cpp]: What is this?

File: toolboxes/mri_core/mri_core_parallel_imaging.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+#pragma once
+
+#include "mri_core_export.h"
+#include "hoNDArray.h"
+
+namespace Gadgetron
+{
+
+template <typename T> 
+class EXPORTMRICORE parallelImaging
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_parallel_imaging.h]: This should not be a class

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+
+template EXPORTMRICORE void cutpad3D(const hoNDArray<float>& data, size_t sizeX, size_t sizeY, size_t sizeZ, hoNDArray<float>& dataCut);
+template EXPORTMRICORE void cutpad3D(const hoNDArray<double>& data, size_t sizeX, size_t sizeY, size_t sizeZ, hoNDArray<double>& dataCut);
+template EXPORTMRICORE void cutpad3D(const hoNDArray< std::complex<float> >& data, size_t sizeX, size_t sizeY, size_t sizeZ, hoNDArray< std::complex<float> >& dataCut);
+template EXPORTMRICORE void cutpad3D(const hoNDArray< std::complex<double> >& data, size_t sizeX, size_t sizeY, size_t sizeZ, hoNDArray< std::complex<double> >& dataCut);
+
+// ------------------------------------------------------------------------
+
+template<typename T> 
+bool sumOverLastDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+    return true;
+}
+
+template EXPORTMRICORE bool sumOverLastDimension(const hoNDArray<float>& x, hoNDArray<float>& r);
+template EXPORTMRICORE bool sumOverLastDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
+template EXPORTMRICORE bool sumOverLastDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
+template EXPORTMRICORE bool sumOverLastDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
+
+template<typename T> 
+bool sumOver1stDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+    {
+        GADGET_ERROR_MSG("Errors in sumOver1stDimension(...) ... ");
+        return false;
+    }
+
+    return true;
+}
+
+template<typename T> 
+bool sumOver2ndDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+    {
+        GADGET_ERROR_MSG("Errors in sumOver2ndDimension(...) ... ");
+        return false;
+    }
+
+    return true;
+}
+
+template<typename T> 
+bool sumOver3rdDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 10]
```diff
@@ -0,0 +1,840 @@
...
+    catch(...)
+    {
+        GADGET_ERROR_MSG("Errors in sumOver3rdDimension(...) ... ");
+        return false;
+    }
+
+    return true;
+}
+
+template<typename T> bool sumOver4thDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 11]
```diff
@@ -0,0 +1,840 @@
...
+    catch(...)
+    {
+        GADGET_ERROR_MSG("Errors in sumOver4thDimension(const hoNDArray<T>& x, hoNDArray<T>& r) ... ");
+        return false;
+    }
+
+    return true;
+}
+
+template<typename T> bool sumOver5thDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Should not be here

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 12]
```diff
@@ -0,0 +1,840 @@
...
+template EXPORTMRICORE bool sumOver4thDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
+
+template EXPORTMRICORE bool sumOver5thDimension(const hoNDArray<float>& x, hoNDArray<float>& r);
+template EXPORTMRICORE bool sumOver5thDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
+template EXPORTMRICORE bool sumOver5thDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
+template EXPORTMRICORE bool sumOver5thDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
+
+// ------------------------------------------------------------------------
+
+inline void multiplyCplx(size_t N, const  std::complex<float> * x, const  std::complex<float> * y,  std::complex<float> * r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Already implemented

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 13]
```diff
@@ -0,0 +1,840 @@
...
+        const float b = a1.imag();
+        const float c = b1.real();
+        const float d = b1.imag();
+
+        reinterpret_cast<float(&)[2]>(r[n])[0] = a*c-b*d;
+        reinterpret_cast<float(&)[2]>(r[n])[1] = a*d+b*c;
+    }
+}
+
+inline void multiplyCplx(size_t N, const  std::complex<double> * x, const  std::complex<double> * y,  std::complex<double> * r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Already implemented

File: toolboxes/mri_core/mri_core_utility.cpp [hunk 14]
```diff
@@ -0,0 +1,840 @@
...
+        const double c = b1.real();
+        const double d = b1.imag();
+
+        reinterpret_cast<double(&)[2]>(r[n])[0] = a*c-b*d;
+        reinterpret_cast<double(&)[2]>(r[n])[1] = a*d+b*c;
+    }
+}
+
+template <typename T> 
+bool multipleMultiply(const hoNDArray<T>& x, const hoNDArray<T>& y, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/mri_core/mri_core_utility.cpp]: Already implemented

**Comment by hansenms** [comment]
One more comment. When you move these functions out of gtPlus, you should delete them from gtPlus. That way there is only one set of parallel imaging functions and there would be some existing tests to verify that they are working. 

And there should be at least one (non gtplus) chain that calls these functions.

**Comment by gtdeveloper5** [comment]
Yes, i already changed the gtplus grappa worker to use the new grappa functions. All tests passed. Will do not cleaning as well.

**Comment by hansenms** [comment]
I am closing this pull request. It is not getting merged in the current form. As discussed:

1. The math/support functions needs to be separated and have unit tests. They also need to be generic without any MRI specific terminology. 
2. The labels need to be a separate pull request. 
3. Then we will deal with what is left. The actual GRAPPA functionality needs to be accompanied by a simple gadget chain that illustrates its use and integration tests.

**Pull request closed without merging**